[wxwidgets] Update to v3.3.0

RippeR37 opened the pull request:
- [x] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md).
- [x] SHA512s are updated for each updated download.
- [ ] The "supports" clause reflects platforms that may be fixed by this new version.
- [ ] Any fixed [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt) entries are removed from that file.
- [ ] Any patches that are no longer applied are deleted from the port's directory.
- [x] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [x] Only one version is added to each modified port's versions file.

----

## Commit description:

This commit updates port of wxWidgets library to v3.3.0. Changes:

* libwebp is was moved from vendored source code to a submodule, fixed by requesting to use system package (via `find_package()`) and adding `libwebp` as dependency,
* `scintilla` and `lexilla` codebases were moved from vendored source-code to submodules, fixed by fetching the repos in portfile and copying their content into the original locations (exact same refs). Rationale below.

Additionally it also fixes building of Edge webview backend and making it link statically when linking everything else statically.

----

### Rationale for chosen solution for `scintilla` and `lexilla` dependencies:

There are few ways to handle this in vcpkg that I can see:

1. Create a huge patch with all missing files and apply it. The biggest problem is that this unfortunately is hard to review and maintain over time so I decided to go with another approach.
2. Make ports for both of these and then depend on them via `find_package()` instead in wxWidget's CMake script. The biggest problems here are that neither of these are compatible with CMake by default, which would be a burden to maintain and update to be able to update wxWidgets, but wxWidget's integration/dependency on these is also complex, which would require complex patch of wxWidget's CMake script to be supported over time. I decided to not pursue this path.
3. Add small patch in wxWidget's CMake script that would fetch the deps on its own when configuring/building wxWidgets. The problem here is that (a) vcpkg disables fetching data from internet in FetchContent() and (b) that this way of fetching non-CMake repos is deprecated and will be removed in the future (see policy CMP0169).
4. Fetch these depdendencies with `vcpkg_from_github()` as separate dependencies and copy into wxWidgets. While not ideal, this has a benefit of being up-front about what is being done, doesn't require changes to wxWidget's build scripts and is easy to maintain over time.

Due to lack of any better way, I decided to go with this solution.

RippeR37 committed:
wxWidgets: Update to v3.3.0

This commit updates port of wxWidgets library to v3.3.0. Changes:

* libwebp is was moved from vendored source code to a submodule, fixed by
  requesting to use system package (via `find_package()`) and adding
  `libwebp` as dependency,
* `scintilla` and `lexilla` codebases were moved from vendored
  source-code to submodules, fixed by fetching the repos in portfile and
  copying their content into the original locations (exact same refs).
  Rationale below.

Additionally it also fixes building of Edge webview backend and making
it link statically when linking everything else statically.

----

Rationale for chosen solution for `scintilla` and `lexilla` dependencies:

There are few ways to handle this in vcpkg that I can see:

1. Create a huge patch with all missing files and apply it. The biggest
   problem is that this unfortunately is hard to review and maintain
   over time so I decided to go with another approach.
2. Make ports for both of these and then depend on them via
   `find_package()` instead in wxWidget's CMake script.
   The biggest problems here are that neither of these are compatible
   with CMake by default, which would be a burden to maintain and update
   to be able to update wxWidgets, but wxWidget's integration/dependency
   on these is also complex, which would require complex patch of
   wxWidget's CMake script to be supported over time. I decided to not
   pursue this path.
3. Add small patch in wxWidget's CMake script that would fetch the deps
   on its own when configuring/building wxWidgets. The problem here is
   that (a) vcpkg disables fetching data from internet in FetchContent()
   and (b) that this way of fetching non-CMake repos is deprecated and
   will be removed in the future (see policy CMP0169).
4. Fetch these depdendencies with `vcpkg_from_github()` as separate
   dependencies and copy into wxWidgets. While not ideal, this has a
   benefit of being up-front about what is being done, doesn't require
   changes to wxWidget's build scripts and is easy to maintain over time.

Due to lack of any better way, I decided to go with this solution.

vicroms commented on ports/wxwidgets/portfile.cmake:
```diff
@@ -14,6 +14,24 @@ vcpkg_from_github(
+    HEAD_REF wx
+)
+file(COPY "${lexilla_SOURCE_PATH}/" DESTINATION "${SOURCE_PATH}/src/stc/lexilla")
+vcpkg_from_github(
+    OUT_SOURCE_PATH scintilla_SOURCE_PATH
+    REPO wxWidgets/scintilla
+    REF "0b90f31ced23241054e8088abb50babe9a44ae67"
+    SHA512 db1f3007f4bd8860fad0817b6cf87980a4b713777025128cf5caea8d6d17b6fafe23fd22ff6886d7d5a420f241d85b7502b85d7e52b4ddb0774edc4b0a0203e7
+    HEAD_REF wx
+)
+file(COPY "${scintilla_SOURCE_PATH}/" DESTINATION "${SOURCE_PATH}/src/stc/scintilla")
```
Depend on port `scintilla` instead of vendoring the dependency.

RippeR37 replied:
Note that this would most likely require notable patches to wxWidget's CMake script, e.g. in:

* https://github.com/wxWidgets/wxWidgets/blob/master/build/cmake/lib/stc/scintilla.cmake
* https://github.com/wxWidgets/wxWidgets/blob/master/build/cmake/lib/stc/CMakeLists.txt
* etc.

(overall non-trivial changes) and also - is it okay thank to land port of 3.3(.x) versions on the `wxwidgets` port (based on the discussion above)? Double-checking as the changes seem to be non-trivial and I'd prefer to avoid spending too much time on this if it would be be blocked anyway later on.

also, is using port of `lexilla` also required, if there isn't one in vcpkg right now (i.e. would I need to create one to land this)? or since its not yet here, it would get a pass, unlike `scintilla` which has a port already.

BillyONeal commented:
The problem appears to be that wxwidgets appears to have effectively forked both of those, so they've created a situation where you can't have wxwidgets and scintilla's libraries on the same system, as they try to stomp on the same symbol names. It seems reasonable to expect someone to want to be able to write an app that uses wxwidgets and also uses scintilla but this vendoring makes that impossible.

Unfortunately the linked examples into wxwidgets' sources don't really help clarify for me; is there a big reason what the rest of wxwidgets needs can't be reduced to a target that could be provided with a find module?

vicroms commented:
@BillyONeal @ras0219-msft @vicroms 

Hi @RippeR37, we discussed this PR during our meeting and have the following comments:

> P.S. There was a discussion whether original wxwidgets port should be updated to v3.3.x since these versions will not guarantee no API/ABI changes between them (e.g. 3.3.x and 3.3.(x+1). As far as I know, this is never guaranteed between port updates, hence why pinning of dependencies exists, but I wanted to bring it up in case this should be a separate port, for example.

* vcpkg doesn't make source compatibility guarantees with regards to port updates,
* the stable release has not received new features since 2022 and the next stable is not scheduled until 2026 at the earliest, and
* all ports depending on `wxwidgets` succeed on a CI test using version 3.3 with no changes

The question becomes: how do we handle `wxwidgets` releases in vcpkg? 

option 1) treat it as a library that breaks some compatibility regularly (e.g.: Boost), or
option 2) treat it as a library that only has new feature updates every couple of years (e.g.: OpenSSL)

We believe that option 1 is a better experience for vcpkg users given that as evidenced by CI, so far, consumers of `wxwidgets` don't require source changes to build with the new version.

In regard to the vendoring of `scintilla` and others, the existing port already vendors those libraries so we don't consider the changes in this PR a new violation of policy and have no reason to block the PR.

Thanks for your work on this PR so far!

RippeR37 commented:
Hi @BillyONeal, @vicroms,

Thank you for detailed answer and in-depth analysis of this! :) My thoughts were exactly the same. 
I'll resume my work on this port then & rebase on top of latest update (to v3.2.8.1). Will ping you once its ready to be reviewed.

I'll also try to do 3.3.0 and then update to 3.3.1 separately in case someone would need a port of 3.3.0 and not 3.3.1 for some reason (so they will be able to pin to that version's port).

RippeR37 committed:
fixup! wxWidgets: Update to v3.3.0

Fixed remaining bugs - remoed empty directory in release-only build, fixed file name to 3.2=>3.3, stopped passing wxUSE_STL and replaced it with other flag for backwards compatibility

RippeR37 committed:
fixup! wxWidgets: Update to v3.3.0

Fix arm(64) building by introducing (temporary) patch to fix wxWidget's cmake script

RippeR37 commented on ports/wxwidgets/fix-cmake-msvc-naming.patch:
```diff
@@ -0,0 +1,13 @@
+ 
+ # Use the MSVC/makefile naming convention, or the configure naming convention,
+ # this is the same check as used in FindwxWidgets.
+-if(WIN32 AND NOT CYGWIN AND NOT MSYS AND NOT CMAKE_CROSSCOMPILING)
++if(MSVC OR (WIN32 AND NOT CYGWIN AND NOT MSYS AND NOT CMAKE_CROSSCOMPILING))
```
FYI I've aligned this implementation with current implementation (also found in v3.3.1: https://github.com/wxWidgets/wxWidgets/blob/v3.3.1/build/cmake/options.cmake#L98). With this, we should be able to simply drop this patch in v3.3.1 port update (and potentially replace with `-DwxBUILD_WIN32_MSVC_NAMING=ON` for ARM, but it might work out of the box either way due to `MSVC`).

RippeR37 committed:
fixup! wxWidgets: Update to v3.3.0

Attempt to fix compilation by adding INTERFACE requirement for C++11 support to library targets

RippeR37 committed:
fixup! wxWidgets: Update to v3.3.0

Another attempt to make wxWidget's target depend on C++11 publicly

RippeR37 committed:
fixup! wxWidgets: Update to v3.3.0

And another attempt...

RippeR37 committed:
fixup! wxWidgets: Update to v3.3.0

Removed attempt to force users of wxwidgets to use C++ >=11, instead fix plplot and wxwidgets/example to use it

RippeR37 committed:
fixup! wxWidgets: Update to v3.3.0

Fix plplot patch, from master to actually used version 5.15.0

RippeR37 committed:
fixup! wxWidgets: Update to v3.3.0

RippeR37 committed:
fixup! wxWidgets: Update to v3.3.0

RippeR37 commented:
Te remaining failures are:

### arm64-windows
```
D:\a\_work\1\s\scripts\azure-pipelines/../ci.feature.baseline.txt:404:1: error: plplot[core]:arm64-windows passed but was marked expected to fail
D:\a\_work\1\s\scripts\azure-pipelines/../ci.feature.baseline.txt:404:1: error: plplot[core,wxwidgets]:arm64-windows passed but was marked expected to fail
```
Should I just update the baseline and make these expected to work?

### x64_uwp
```
D:\a\_work\1\s\scripts\azure-pipelines/../ci.feature.baseline.txt:405:1: error: plplot[core]:x64-uwp was unexpectedly a cascading failure because the following dependencies are unavailable: cairo[core]:x64-uwp only supports !xbox & !uwp, fontconfig[core]:x64-uwp only supports !uwp, fribidi[core]:x64-uwp only supports !uwp, glib[core]:x64-uwp only supports !uwp & !xbox, harfbuzz[core]:x64-uwp only supports !uwp
D:\a\_work\1\s\scripts\azure-pipelines/../ci.feature.baseline.txt:405:1: note: consider changing this to =cascade instead
D:\a\_work\1\s\scripts\azure-pipelines/../ci.feature.baseline.txt:405:1: error: plplot[core,wxwidgets]:x64-uwp was unexpectedly a cascading failure because the following dependencies are unavailable: cairo[core]:x64-uwp only supports !xbox & !uwp, fontconfig[core]:x64-uwp only supports !uwp, fribidi[core]:x64-uwp only supports !uwp, glib[core]:x64-uwp only supports !uwp & !xbox, harfbuzz[core]:x64-uwp only supports !uwp, opengl[core]:x64-uwp only supports !uwp & !xbox, wxwidgets[core]:x64-uwp only supports !uwp & !xbox
D:\a\_work\1\s\scripts\azure-pipelines/../ci.feature.baseline.txt:405:1: note: consider changing this to =cascade instead
D:\a\_work\1\s\scripts\azure-pipelines/../ci.feature.baseline.txt:405:1: error: plplot[core,x11]:x64-uwp was unexpectedly a cascading failure because the following dependencies are unavailable: cairo[core]:x64-uwp only supports !xbox & !uwp, cairo[x11]:x64-uwp only supports !windows, fontconfig[core]:x64-uwp only supports !uwp, fribidi[core]:x64-uwp only supports !uwp, glib[core]:x64-uwp only supports !uwp & !xbox, harfbuzz[core]:x64-uwp only supports !uwp, libxslt[core]:x64-uwp only supports !uwp, xorg-macros[core]:x64-uwp only supports !uwp
D:\a\_work\1\s\scripts\azure-pipelines/../ci.feature.baseline.txt:405:1: note: consider changing this to =cascade instead
D:\a\_work\1\s\scripts\azure-pipelines/../ci.feature.baseline.txt:405:1: error: plplot[core,wxwidgets,x11]:x64-uwp was unexpectedly a cascading failure because the following dependencies are unavailable: cairo[core]:x64-uwp only supports !xbox & !uwp, cairo[x11]:x64-uwp only supports !windows, fontconfig[core]:x64-uwp only supports !uwp, fribidi[core]:x64-uwp only supports !uwp, glib[core]:x64-uwp only supports !uwp & !xbox, harfbuzz[core]:x64-uwp only supports !uwp, libxslt[core]:x64-uwp only supports !uwp, opengl[core]:x64-uwp only supports !uwp & !xbox, wxwidgets[core]:x64-uwp only supports !uwp & !xbox, xorg-macros[core]:x64-uwp only supports !uwp
D:\a\_work\1\s\scripts\azure-pipelines/../ci.feature.baseline.txt:405:1: note: consider changing this to =cascade instead```

No idea how `x64_uwp` was supposed to work before if these libraries (cairo, ...) don't support `uwp` themselves.

Other than that I thinks its ready to be reviewed.

RippeR37 committed:
fixup! wxWidgets: Update to v3.3.0

Update CI baseline

RippeR37 committed:
fixup! wxWidgets: Update to v3.3.0

Update CI baseline

RippeR37 committed:
fixup! wxWidgets: Update to v3.3.0

Update CI baseline

RippeR37 committed:
[empty commit] Re-trigger checks, one failed randomly with HTTP 503 on vcpkg init

RippeR37 committed:
Re-trigger CI

RippeR37 commented:
@vicroms
Whew, finally got a fully passing CI. Should be good I hope. Once this lands, I'll open up new PR with 3.3.1 port update.

Thank you for your help here! :)

BillyONeal commented:
Thanks for the update!

dg0yt commented:
> AFAIU 3.3 may still introduce too many / too frequent changes to be a good choice for this registry. This affects users and CI. Pinning exists, but switching to 3.3 would prevent integrating further updates to the 3.2 series.

"affects users": #46736
